### ISSUE TITLE
Bluetooth headset: media key recognized in shortcuts, but play button not working

### ISSUE THREAD

**Issue by Alexey-Yakovenko** [Alexey-Yakovenko/deadbeef#1180]
Original [issue 1279](https://code.google.com/p/ddb/issues/detail?id=1279) created by Alexey-Yakovenko on 2015-02-22T15:25:48.000Z:

<b>What steps will reproduce the problem?</b>
<b>Какие шаги приводят к воспроизведению</b>
<b>проблемы?</b>
1. pair a BT headset in a2dp profile
2. define the play/pause toggle shortcut (global) in the shortcut manager
3. try controlling the playback with the media keys on the headset

<b>What is the expected output? What do you see instead?</b>
<b>Какой ожидаемый вывод? Что вы видите</b>
<b>вместо него?</b>
The shortcut is correctly set to XF86AudioPlay when defining it - yet the play/pause key is without function. Same key (with the same keycode) on the keyboard works fine. Other keys (prev/next) on the BT headset work fine.

<b>What version of the product are you using? On what operating system and CPU</b>
<b>architecture?</b>
<b>Какую версию продукта вы используете? На</b>
<b>какой операционной системе и</b>
<b>архитектуре CPU?</b>
deadbeef-devel, arch x64

<b>How did you install the product?</b>
<b>Как вы установили продукт?</b>
deadbeef-git from AUR

<b>Please provide any additional information below.</b>
<b>Пожалуйста, предоставьте любую</b>
<b>дополнительную информацию ниже.</b>
It might be somehow related, so I'll put this here: I had a similar problem with this headset on android, I could only unpause playback, but not pause it. According to android logs, the headset was sending a &quot;Play&quot; signal, not &quot;PlayPause&quot; - but interestingly only on android 4.3. With android 4.4 the headset is sending &quot;PlayPause&quot; signal as expected.

But here, it's sending the Play signal, which is for some reason working in the hotkey manager, but is not caught by the actual playback control function?

**Comment by Soukyuu** [comment]
Might be a stupid question, but how do you make messages like these actually appear while the program is running?
```
trace ("init: key %s code %x\n", name, keys[i].keysym);
```
I'm trying to find out the difference and why it's recognized correctly on the config dialog but not while actually using ddb.

**Comment by cboxdoerfer** [comment]
At the beginning of the resp. source file there are two definitions of trace, just uncomment the one which calls fprintf and comment the other.

**Comment by Soukyuu** [comment]
Ok, thanks. Here is the difference, maybe it could point you to the right direction.

This is the output when I press the X86AudioPlay key on my keyboard:
```
hotkeys: key 172 mods 0 (10)
filtered state=0
matches to commands[30]!
```
And this is what happens if I do the same on my headset:
```
hotkeys: keysym 0x1008FF14 mapped to 0x1008FF14
hotkeys: command find keycode 66 mods 4
hotkeys: command open_files keycode 6f mods 4
hotkeys: command quit keycode 71 mods 4
hotkeys: command new_playlist keycode 6e mods 4
hotkeys: command select_all keycode 61 mods 4
hotkeys: command deselect_all keycode ff1b mods 0
hotkeys: command toggle_stop_after_current keycode 6d mods 4
hotkeys: command jump_to_current_track keycode 6a mods 4
hotkeys: command help keycode ffbe mods 0
hotkeys: command remove_from_playlist keycode ffff mods 0
hotkeys: command remove_current_playlist keycode 77 mods 4
hotkeys: command play keycode ff0d mods 0
hotkeys: command preferences keycode 70 mods 4
hotkeys: command playlist1 keycode 31 mods 8
hotkeys: command playlist2 keycode 32 mods 8
hotkeys: command playlist3 keycode 33 mods 8
hotkeys: command playlist4 keycode 34 mods 8
hotkeys: command playlist5 keycode 35 mods 8
hotkeys: command playlist6 keycode 36 mods 8
hotkeys: command playlist7 keycode 37 mods 8
hotkeys: command playlist8 keycode 38 mods 8
hotkeys: command playlist9 keycode 39 mods 8
hotkeys: command playlist10 keycode 30 mods 8
hotkeys: command prev keycode 7a mods 0
hotkeys: command play keycode 78 mods 0
hotkeys: command toggle_pause keycode 63 mods 0
hotkeys: command stop keycode 76 mods 0
hotkeys: command next keycode 62 mods 0
hotkeys: command playback_random keycode 6e mods 0
hotkeys: command toggle_stop_after_album keycode 6b mods 4
hotkeys: command play_pause keycode 1008ff14 mods 0
hotkeys: command next keycode 1008ff17 mods 0
hotkeys: command prev keycode 1008ff16 mods 0
hotkeys: command stop keycode 1008ff15 mods 0
hotkeys: command toggle_eq keycode 65 mods 4
hotkeys: command load_playlist keycode 6c mods 5
hotkeys: command add_files keycode 6c mods 4
```
After that, the headset key does not trigger any messages anymore - it's dead. Restarting ddb returns to the same behavior. 

Oh and as another proof it's not my system: I can actually control mpv playback with my headset's play button.

**Comment by Soukyuu** [comment]
Ok, I've got it. The headset key fails the `isglobal` check:
```
if (commands[i].keycode == keycode && commands[i].modifier == mods && commands[i].isglobal == isglobal)
```
gives
```
hotkeys: 1008ff14 == 1008ff14, 0 == 0, 1 != 0
```
So:
- if ddb is focused: headset sends a local "play" action, fails ddb's `isglobal` check.
- if ddb is not focused: headset's "play" action does not seem to reach ddb at all.

**Comment by Alexey-Yakovenko** [comment]
@Soukyuu looks like the key code is not mapped correctly. Please try the following:

run terminal, then run ```xev``` in it, then start pressing buttons on your BT headset (make sure it's paired).

What does it print?

**Comment by Alexey-Yakovenko** [comment]
BTW, does it work if you make a non-global hotkey?

**Comment by Soukyuu** [comment]
the "play" key gives:
```
KeyPress event, serial 41, synthetic NO, window 0x5200001,
    root 0x263, subw 0x0, time 5609283, (250,278), root:(2721,768),
    state 0x10, keycode 208 (keysym 0x1008ff14, XF86AudioPlay), same_screen YES,
    XKeysymToKeycode returns keycode: 172
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 41, synthetic NO, window 0x5200001,
    root 0x263, subw 0x0, time 5609294, (250,278), root:(2721,768),
    state 0x10, keycode 208 (keysym 0x1008ff14, XF86AudioPlay), same_screen YES,
    XKeysymToKeycode returns keycode: 172
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```
prev/next don't give any output. This is when the xev window is focused. If the xev window is not focused, no keys on the headset give any output.

For comparison: the keyboard play key:
```
KeyPress event, serial 68, synthetic NO, window 0x5200001,
    root 0x263, subw 0x0, time 5822932, (102,350), root:(2573,840),
    state 0x10, keycode 172 (keysym 0x1008ff14, XF86AudioPlay), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 68, synthetic NO, window 0x5200001,
    root 0x263, subw 0x0, time 5823080, (102,350), root:(2573,840),
    state 0x10, keycode 172 (keysym 0x1008ff14, XF86AudioPlay), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```
xev does list the keypress/release events for prev/next on the keyboard, contrary to the headset ones. 

And yes, making a non-global hotkey works (as long as deadbeef is focused)

**Comment by Alexey-Yakovenko** [comment]
Thanks

**Comment by Soukyuu** [comment]
As discussed on irc, this doesn't seem to be a bug with deadbeef - the keypress never reaches it.
For now, a workaround is possible by using xbindkeys.

**Comment by Alexey-Yakovenko** [comment]
Closed because no way to fix this on deadbeef side.
